Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h

```diff
@@ -114,11 +115,13 @@ class PDFPluginBase : public ThreadSafeRefCounted<PDFPluginBase>, public WebCore
     virtual bool handleWheelEvent(const WebWheelEvent&) = 0;
     virtual bool handleMouseEnterEvent(const WebMouseEvent&) = 0;
     virtual bool handleMouseLeaveEvent(const WebMouseEvent&) = 0;
-    virtual bool handleContextMenuEvent(const WebMouseEvent&) = 0;
     virtual bool handleKeyboardEvent(const WebKeyboardEvent&) = 0;
     virtual bool handleEditingCommand(StringView commandName) = 0;
     virtual bool isEditingCommandEnabled(StringView commandName) = 0;
 
+    bool handleContextMenuEvent(const WebMouseEvent&);
+    virtual NSMenu* createContextMenu(const WebMouseEvent&, WebCore::UserInterfaceLayoutDirection) = 0;
```

Comment:
I think at the PDFPluginBase level we should be using platform-agnostic PDFContextMenu, not NSMenu, and PDFPlugin (the old one) should do the conversion internally.

Reply:
To be even more clear: I don't think UnifiedPDFPlugin should ever have an NSMenu involved (unless there's some good reason I'm not seeing).

Reply:
If you do happen to keep `NSMenu* createContextMenu()`, you need to get the refcounting correct. The "create" in the name impies that it returns a retained value, which it does not. It would be clearer to return a `RetainPtr<NSMenu>`.

Reply:
How about put the responsibility of returning a PDFContextMenu on the plugin implementation? That way plugin base stays platform agnostic, PDFPlugin can go from NSMenu to PDFContextMenu internally, and UniversalPDFPlugin can also represent context menus however it wants internally as long as it is providing the plugin base a PDFContextMenu object. (Basically what Tim suggested earlier)